### PR TITLE
Add ProjectLoggerContext

### PR DESCRIPTION
**Pull request by davkean**
This lets a higher level component prepend additional information to a logged message, meaning that less information needs to be passed down the lower components.

This simplifies logger significantly in the language service; the LanguageServiceHost begins a context and then passes it down to the handlers which simply log just what they are doing and don't need to include information about the configured project version or configuration, etc.

Note this includes: https://github.com/dotnet/project-system/pull/2142,k so just look at last 3 commits.

### CODE REVIEW

**Comment by davkean** [comment]
tag @dotnet/project-system

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Logging/FormatArray.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+namespace Microsoft.VisualStudio.ProjectSystem.Logging
+{
+    /// <summary>
+    ///     Provides an object that allows a common way to format a set of objects, only allocating if needed.
+    /// </summary>
+    internal struct FormatArray
+    {
+        // Sentinel fixed-length arrays eliminate the need for a "count" field keeping this
+        // struct down to just 4 fields. These are only used for their "Length" property,
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Logging/FormatArray.cs]: Five fields?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Logging/ProjectLoggerContext.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        ///     <paramref name="format"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="FormatException">
+        ///     The format specification in <paramref name="format"/> is invalid.
+        /// </exception>
+        public void WriteLine(string format, object argument)
+        {
+            if (_logger.IsEnabled)
+            {
+                _logger.WriteLine(_formatArray.Text + " " + format, argument);
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Logging/ProjectLoggerContext.cs]: Could this run into problems if `_formatArray.Text` returns a string that itself contains formatting placeholders? And if so, do we care?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Logging/ProjectOutputWindowProjectLogger.cs [hunk 3]
```diff
@@ -30,38 +29,52 @@ public ProjectOutputWindowProjectLogger(IProjectThreadingService threadingServic
...
+            WriteLine(new FormatArray(format, argument1, argument2, argument3));
+        }
+
+        public void WriteLine(string format, params object[] arguments)
+        {
+            WriteLine(new FormatArray(format, arguments));
         }
 
-        private void WriteLineCore(string text)
+        private void WriteLine(FormatArray formatArray)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Logging/ProjectOutputWindowProjectLogger.cs]: Is there really no better API to write to the output pane with?

**Pull request closed without merging**